src/Functions/MultiMatchAllIndicesImpl.h

```diff
@@ -120,6 +121,128 @@ struct MultiMatchAllIndicesImpl
             offset = haystack_offsets[i];
         }
 #else
+        (void)haystack_data;
+        (void)haystack_offsets;
+        (void)needles_arr;
+        (void)res;
+        (void)offsets;
+        (void)edit_distance;
+        (void)max_hyperscan_regexp_length;
+        (void)max_hyperscan_regexp_total_length;
+        throw Exception(
+            "multi-search all indices is not implemented when vectorscan is off",
+            ErrorCodes::NOT_IMPLEMENTED);
+#endif // USE_VECTORSCAN
+    }
+
+    static void vectorVector(
+        const ColumnString::Chars & haystack_data,
+        const ColumnString::Offsets & haystack_offsets,
+        const IColumn & needles_data,
+        const ColumnArray::Offsets & needles_offsets,
+        PaddedPODArray<ResultType> & res,
+        PaddedPODArray<UInt64> & offsets,
+        bool allow_hyperscan,
+        size_t max_hyperscan_regexp_length,
+        size_t max_hyperscan_regexp_total_length)
+    {
+        vectorVector(haystack_data, haystack_offsets, needles_data, needles_offsets, res, offsets, std::nullopt, allow_hyperscan, max_hyperscan_regexp_length, max_hyperscan_regexp_total_length);
+    }
+
+    static void vectorVector(
+        const ColumnString::Chars & haystack_data,
+        const ColumnString::Offsets & haystack_offsets,
+        const IColumn & needles_data,
+        const ColumnArray::Offsets & needles_offsets,
+        PaddedPODArray<ResultType> & res,
+        PaddedPODArray<UInt64> & offsets,
+        std::optional<UInt32> edit_distance,
+        bool allow_hyperscan,
+        size_t max_hyperscan_regexp_length,
+        size_t max_hyperscan_regexp_total_length)
+    {
+        if (!allow_hyperscan)
+            throw Exception(ErrorCodes::FUNCTION_NOT_ALLOWED, "Hyperscan functions are disabled, because setting 'allow_hyperscan' is set to 0");
+#if USE_VECTORSCAN
+        offsets.resize(haystack_offsets.size());
+        size_t prev_haystack_offset = 0;
+
+        const ColumnString * needles_data_string = checkAndGetColumn<ColumnString>(&needles_data);
+        const ColumnString::Offsets & needles_data_string_offsets = needles_data_string->getOffsets();
+        const ColumnString::Chars & needles_data_string_chars = needles_data_string->getChars();
+
+        std::vector<std::string_view> needles;
+
+        size_t prev_needles_offsets_offset = 0;
+        size_t prev_needles_data_offset = 0;
+
+        for (size_t i = 0; i < haystack_offsets.size(); ++i)
+        {
+            needles.reserve(needles_offsets[i] - prev_needles_offsets_offset);
+
+            for (size_t j = prev_needles_offsets_offset; j < needles_offsets[i]; ++j)
+            {
+                const auto * p = reinterpret_cast<const char *>(needles_data_string_chars.data()) + prev_needles_data_offset;
```

Comment:
Strings are no longer copied out of the array. Instead, I added logic to reference them by index. I also looked at `getDataAt()` (which you recommended) but it works only with fixed-width types, not strings.

Reply:
> but it works only with fixed-width types, not strings.

That's not true, `ColumnString` has it's own implementation of `getDataAt()`
https://github.com/ClickHouse/ClickHouse/blob/74d02aeca759646564540843016539195469b1c3/src/Columns/ColumnString.h#L105-L109

Reply:
You are right, I looked in the wrong file. Your proposal simplifies the code a lot.